SQL Deadlock 7.3

```Error updating fields for: DocRef{type='Index', uuid='xxxxxxx', name='xxxxxxx'} SQL [insert into `index_field` (`fk_index_field_source_id`, `type`, `name`, `analyzer`, `indexed`, `stored`, `term_positions`, `case_sensitive`) values (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?) on duplicate key update `index_field`.`fk_index_field_source_id` = ?]; Deadlock found when trying to get lock; try restarting transaction 
org.jooq.exception.DataAccessException: SQL [insert into `index_field` (`fk_index_field_source_id`, `type`, `name`, `analyzer`, `indexed`, `stored`, `term_positions`, `case_sensitive`) values (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?) on duplicate key update `index_field`.`fk_index_field_source_id` = ?]; Deadlock found when trying to get lock; try restarting transaction
	at org.jooq_3.18.7.MYSQL.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:3470)
	at org.jooq.impl.Tools.translate(Tools.java:3458)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:802)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:360)
	at org.jooq.impl.AbstractDelegatingQuery.execute(AbstractDelegatingQuery.java:115)
	at stroom.index.impl.db.IndexFieldDaoImpl.lambda$addFields$2(IndexFieldDaoImpl.java:106)
	at stroom.db.util.JooqUtil.context(JooqUtil.java:118)
	at stroom.index.impl.db.IndexFieldDaoImpl.addFields(IndexFieldDaoImpl.java:84)
	at stroom.index.impl.IndexFieldServiceImpl.addFields(IndexFieldServiceImpl.java:47)
	at stroom.index.impl.DynamicIndexingFilter.flushFields(DynamicIndexingFilter.java:156)
	at stroom.index.impl.DynamicIndexingFilter.endProcessing(DynamicIndexingFilter.java:149)
	at stroom.pipeline.filter.XMLFilterFork.endProcessing(XMLFilterFork.java:89)
	at stroom.pipeline.filter.AbstractXMLFilter.endProcessing(AbstractXMLFilter.java:52)
	at stroom.pipeline.filter.AbstractXMLFilter.endProcessing(AbstractXMLFilter.java:52)
	at stroom.pipeline.filter.XsltFilter.endProcessing(XsltFilter.java:232)
	at stroom.pipeline.filter.AbstractXMLFilter.endProcessing(AbstractXMLFilter.java:52)
	at stroom.pipeline.filter.AbstractXMLFilter.endProcessing(AbstractXMLFilter.java:52)
	at stroom.pipeline.filter.SplitFilter.endProcessing(SplitFilter.java:146)
	at stroom.pipeline.filter.AbstractXMLFilter.endProcessing(AbstractXMLFilter.java:52)
	at stroom.pipeline.parser.AbstractParser.endProcessing(AbstractParser.java:136)
	at stroom.pipeline.reader.AbstractIOElement.endProcessing(AbstractIOElement.java:195)
	at stroom.pipeline.factory.PipelineImpl.internalEndProcessing(PipelineImpl.java:75)
	at stroom.pipeline.factory.PipelineImpl.endProcessing(PipelineImpl.java:55)
	at stroom.core.dataprocess.AbstractProcessorTaskExecutor.processNestedStreams(AbstractProcessorTaskExecutor.java:487)
	at stroom.core.dataprocess.AbstractProcessorTaskExecutor.process(AbstractProcessorTaskExecutor.java:339)
	at stroom.core.dataprocess.AbstractProcessorTaskExecutor.exec(AbstractProcessorTaskExecutor.java:238)
	at stroom.processor.impl.DataProcessorTaskHandler.exec(DataProcessorTaskHandler.java:164)
	at stroom.processor.impl.DataProcessorTaskHandler.lambda$exec$0(DataProcessorTaskHandler.java:92)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$2(TaskContextFactoryImpl.java:253)
	at stroom.util.logging.LocationAwareLambdaLogger.logDurationIfDebugEnabled(LocationAwareLambdaLogger.java:307)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$4(TaskContextFactoryImpl.java:253)
	at stroom.util.pipeline.scope.PipelineScopeRunnable.scopeResult(PipelineScopeRunnable.java:39)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$5(TaskContextFactoryImpl.java:250)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$6(TaskContextFactoryImpl.java:262)
	at stroom.security.impl.SecurityContextImpl.useAsReadResult(SecurityContextImpl.java:390)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:50)
	at jdk.proxy2/jdk.proxy2.$Proxy120.useAsReadResult(Unknown Source)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$7(TaskContextFactoryImpl.java:262)
	at stroom.security.impl.SecurityContextImpl.asUserResult(SecurityContextImpl.java:322)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:50)
	at jdk.proxy2/jdk.proxy2.$Proxy120.asUserResult(Unknown Source)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$11(TaskContextFactoryImpl.java:260)
	at stroom.processor.impl.DataProcessorTaskHandler.lambda$exec$1(DataProcessorTaskHandler.java:92)
	at stroom.security.impl.SecurityContextImpl.useAsReadResult(SecurityContextImpl.java:390)
	at stroom.processor.impl.DataProcessorTaskHandler.lambda$exec$2(DataProcessorTaskHandler.java:87)
	at stroom.security.impl.SecurityContextImpl.asUserResult(SecurityContextImpl.java:322)
	at stroom.processor.impl.DataProcessorTaskHandler.exec(DataProcessorTaskHandler.java:87)
	at stroom.processor.impl.DataProcessorTaskFactory$RunnableFactoryImpl.lambda$create$0(DataProcessorTaskFactory.java:166)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.mysql.cj.jdbc.exceptions.MySQLTransactionRollbackException: Deadlock found when trying to get lock; try restarting transaction
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:124)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:916)
	at com.mysql.cj.jdbc.ClientPreparedStatement.execute(ClientPreparedStatement.java:354)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:445)
	at org.jooq.impl.AbstractDMLQuery.execute(AbstractDMLQuery.java:1024)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:346)
	... 52 common frames omitted```



```
------------------------
LATEST DETECTED DEADLOCK
------------------------
2024-04-09 13:30:42 139696266131200
*** (1) TRANSACTION:
TRANSACTION 2720357778, ACTIVE 0 sec inserting
mysql tables in use 1, locked 1
LOCK WAIT 7 lock struct(s), heap size 1128, 5 row lock(s)
MySQL thread id 27355, OS thread handle 139696497997568, query id 479742158 host IP ref_aceuser update
insert into `index_field` (`fk_index_field_source_id`, `type`, `name`, `analyzer`, `indexed`, `stored`, `term_positions`, `case_sensitive`) values (5, 0, 'EventId', 'Keyword', 1, 1, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric',
 1, 0, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 0, 'StreamId', 'Keyword', 1, 1, 0, 0), (5, 7, 'FIELD2', 'Alpha numeric', 1, 0, 0, 0), (5, 6, 'Time', 'Keyword', 1, 0, 0, 0), (5, 7, 'FIELD3', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD4', 'K
eyword', 1, 0, 0, 0), (5, 7, 'FIELD5', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD6', 'Alpha numeric', 1, 0, 0, 0) on duplicate key update `index_field`.`fk_index_field_source_id` = 5

*** (1) HOLDS THE LOCK(S):
RECORD LOCKS space id 86052 page no 4 n bits 240 index PRIMARY of table `ace`.`index_field` trx id 2720357778 lock_mode X
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;


*** (1) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 86052 page no 4 n bits 240 index PRIMARY of table `ace`.`index_field` trx id 2720357778 lock_mode X insert intention waiting
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;


*** (2) TRANSACTION:
TRANSACTION 2720357777, ACTIVE 0 sec inserting
mysql tables in use 1, locked 1
LOCK WAIT 8 lock struct(s), heap size 1128, 7 row lock(s)
MySQL thread id 27320, OS thread handle 139675585947392, query id 479742157 host ip ref_aceuser update
insert into `index_field` (`fk_index_field_source_id`, `type`, `name`, `analyzer`, `indexed`, `stored`, `term_positions`, `case_sensitive`) values (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 0, 'EventId', 'Keyword', 1, 1, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric
', 1, 0, 0, 0), (5, 0, 'StreamId', 'Keyword', 1, 1, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0), (5, 6, 'Time', 'Keyword', 1, 0, 0, 0), (5, 7, 'FIELD', 'Alpha numeric', 1, 0, 0, 0), (5, 7, 'FIELD', 'Keyword', 1, 0, 0, 0), (5, 7, 'FIELD1', 'Alpha numeric', 1, 0, 0, 0) on duplicate key update
`index_field`.`fk_index_field_source_id` = 5

*** (2) HOLDS THE LOCK(S):
RECORD LOCKS space id 86052 page no 4 n bits 240 index PRIMARY of table `ace`.`index_field` trx id 2720357777 lock_mode X
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;


*** (2) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 86052 page no 4 n bits 240 index PRIMARY of table `ace`.`index_field` trx id 2720357777 lock_mode X insert intention waiting
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;

*** WE ROLL BACK TRANSACTION (1)

```

--- a/stroom-core-shared/src/main/java/stroom/index/shared/IndexFieldImpl.java
+++ b/stroom-core-shared/src/main/java/stroom/index/shared/IndexFieldImpl.java
@@ -133,6 +133,10 @@ public class IndexFieldImpl implements IndexField {
         return new Builder(this);
     }
 
+
+    // --------------------------------------------------------------------------------
+
+
     public static final class Builder {
 
         private String fldName;

--- a/stroom-core-shared/src/main/java/stroom/index/shared/LuceneIndexField.java
+++ b/stroom-core-shared/src/main/java/stroom/index/shared/LuceneIndexField.java
@@ -247,7 +247,6 @@ public class LuceneIndexField implements IndexField {
         return termPositions;
     }
 
-
     @Override
     @JsonIgnore
     public String getDisplayValue() {
@@ -294,6 +293,10 @@ public class LuceneIndexField implements IndexField {
     public Builder copy() {
         return new Builder(this);
     }
+
+
+    // --------------------------------------------------------------------------------
+
 
     public static final class Builder {
 

--- a/stroom-index/stroom-index-impl-db/src/main/java/stroom/index/impl/db/IndexFieldDaoImpl.java
+++ b/stroom-index/stroom-index-impl-db/src/main/java/stroom/index/impl/db/IndexFieldDaoImpl.java
@@ -29,6 +29,7 @@ import stroom.util.shared.ResultPage;
 
 import jakarta.inject.Inject;
 import org.jooq.Condition;
+import org.jooq.DSLContext;
 import org.jooq.Field;
 import org.jooq.impl.DSL;
 
@@ -49,40 +50,64 @@ public class IndexFieldDaoImpl implements IndexFieldDao {
         this.queryDatasourceDbConnProvider = queryDatasourceDbConnProvider;
     }
 
-    private int getOrCreateFieldSource(final DocRef docRef) {
-        Optional<Integer> optional = getFieldSource(docRef);
-        if (optional.isEmpty()) {
-            createFieldSource(docRef);
-            optional = getFieldSource(docRef);
+    private void ensureFieldSource(final DocRef docRef) {
+        JooqUtil.context(queryDatasourceDbConnProvider, context -> {
+            Optional<Integer> optional = getFieldSource(context, docRef, false);
+            if (optional.isEmpty()) {
+                createFieldSource(context, docRef);
+            }
+        });
+    }
+
+    private Optional<Integer> getFieldSource(final DocRef docRef,
+                                             final boolean lockFieldSource) {
+
+        return JooqUtil.contextResult(queryDatasourceDbConnProvider, context ->
+                getFieldSource(context, docRef, lockFieldSource));
+    }
+
+    private Optional<Integer> getFieldSource(final DSLContext context,
+                                             final DocRef docRef,
+                                             final boolean lockFieldSource) {
+        var c = context
+                .select(INDEX_FIELD_SOURCE.ID)
+                .from(INDEX_FIELD_SOURCE)
+                .where(INDEX_FIELD_SOURCE.TYPE.eq(docRef.getType()))
+                .and(INDEX_FIELD_SOURCE.UUID.eq(docRef.getUuid()));
+
+        if (lockFieldSource) {
+            return c.forUpdate()
+                    .fetchOptional(INDEX_FIELD_SOURCE.ID);
+        } else {
+            return c.fetchOptional(INDEX_FIELD_SOURCE.ID);
         }
-        return optional.orElseThrow();
     }
 
-    private Optional<Integer> getFieldSource(final DocRef docRef) {
-        return JooqUtil
-                .contextResult(queryDatasourceDbConnProvider, context -> context
-                        .select(INDEX_FIELD_SOURCE.ID)
-                        .from(INDEX_FIELD_SOURCE)
-                        .where(INDEX_FIELD_SOURCE.TYPE.eq(docRef.getType()))
-                        .and(INDEX_FIELD_SOURCE.UUID.eq(docRef.getUuid()))
-                        .fetchOptional(INDEX_FIELD_SOURCE.ID));
-    }
-
-    private void createFieldSource(final DocRef docRef) {
-        JooqUtil.context(queryDatasourceDbConnProvider, context -> context
+    private void createFieldSource(final DSLContext context, final DocRef docRef) {
+        // TODO Consider using JooqUtil.tryCreate() as onDuplicateKeyIgnore will
+        //  ignore any error, not just dup keys
+        context
                 .insertInto(INDEX_FIELD_SOURCE)
                 .set(INDEX_FIELD_SOURCE.TYPE, docRef.getType())
                 .set(INDEX_FIELD_SOURCE.UUID, docRef.getUuid())
                 .set(INDEX_FIELD_SOURCE.NAME, docRef.getName())
                 .onDuplicateKeyIgnore()
-                .execute());
+                .execute();
     }
 
     @Override
     public void addFields(final DocRef docRef, final Collection<IndexField> fields) {
-        final int fieldSourceId = getOrCreateFieldSource(docRef);
-        JooqUtil.context(queryDatasourceDbConnProvider, context -> {
-            var c = context.insertInto(INDEX_FIELD,
+        // Do this outside the txn so other threads can see it asap
+        ensureFieldSource(docRef);
+
+        JooqUtil.transaction(queryDatasourceDbConnProvider, txnContext -> {
+            // Get a record lock on the field source, so we are the only thread
+            // that can mutate the index fields for that source, else we can get a deadlock.
+            final int fieldSourceId = getFieldSource(txnContext, docRef, true)
+                    .orElseThrow(() -> new RuntimeException("No field source found for " + docRef));
+
+            // Insert any new fields under lock
+            var c = txnContext.insertInto(INDEX_FIELD,
                     INDEX_FIELD.FK_INDEX_FIELD_SOURCE_ID,
                     INDEX_FIELD.TYPE,
                     INDEX_FIELD.NAME,
@@ -91,6 +116,7 @@ public class IndexFieldDaoImpl implements IndexFieldDao {
                     INDEX_FIELD.STORED,
                     INDEX_FIELD.TERM_POSITIONS,
                     INDEX_FIELD.CASE_SENSITIVE);
+
             for (final IndexField field : fields) {
                 c = c.values(fieldSourceId,
                         (byte) field.getFldType().getIndex(),
@@ -101,6 +127,7 @@ public class IndexFieldDaoImpl implements IndexFieldDao {
                         field.isTermPositions(),
                         field.isCaseSensitive());
             }
+            // Effectively ignore existing fields
             c.onDuplicateKeyUpdate()
                     .set(INDEX_FIELD.FK_INDEX_FIELD_SOURCE_ID, fieldSourceId)
                     .execute();
@@ -109,7 +136,8 @@ public class IndexFieldDaoImpl implements IndexFieldDao {
 
     @Override
     public ResultPage<IndexField> findFields(final FindIndexFieldCriteria criteria) {
-        final Optional<Integer> optional = getFieldSource(criteria.getDataSourceRef());
+        final Optional<Integer> optional = getFieldSource(criteria.getDataSourceRef(), false);
+
         if (optional.isEmpty()) {
             return ResultPage.createCriterialBasedList(Collections.emptyList(), criteria);
         }

--- a/stroom-index/stroom-index-impl-db/src/test/java/stroom/index/impl/db/TestIndexFieldDaoImpl.java
+++ b/stroom-index/stroom-index-impl-db/src/test/java/stroom/index/impl/db/TestIndexFieldDaoImpl.java
@@ -1,0 +1,226 @@
+package stroom.index.impl.db;
+
+import stroom.datasource.api.v2.AnalyzerType;
+import stroom.datasource.api.v2.FieldType;
+import stroom.datasource.api.v2.FindIndexFieldCriteria;
+import stroom.datasource.api.v2.IndexField;
+import stroom.db.util.JooqUtil;
+import stroom.docref.DocRef;
+import stroom.index.impl.IndexFieldDao;
+import stroom.index.shared.IndexFieldImpl;
+import stroom.index.shared.LuceneIndexDoc;
+import stroom.util.concurrent.ThreadUtil;
+import stroom.util.exception.ThrowingRunnable;
+import stroom.util.logging.LambdaLogger;
+import stroom.util.logging.LambdaLoggerFactory;
+import stroom.util.shared.PageRequest;
+import stroom.util.shared.ResultPage;
+
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import jakarta.inject.Inject;
+import org.jooq.Record1;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static stroom.db.util.JooqUtil.count;
+import static stroom.index.impl.db.jooq.tables.IndexField.INDEX_FIELD;
+import static stroom.index.impl.db.jooq.tables.IndexFieldSource.INDEX_FIELD_SOURCE;
+
+class TestIndexFieldDaoImpl {
+
+    private static final LambdaLogger LOGGER = LambdaLoggerFactory.getLogger(TestIndexFieldDaoImpl.class);
+
+    private static final DocRef DOC_REF_1 = DocRef.builder()
+            .uuid("a8c078d8-b2e7-41e6-b254-d4870ab75ab8")
+            .type(LuceneIndexDoc.DOCUMENT_TYPE)
+            .name("foo")
+            .build();
+
+    private static final DocRef DOC_REF_2 = DocRef.builder()
+            .uuid("05da4d75-28d4-48ad-9195-9302cfacbc1c")
+            .type(LuceneIndexDoc.DOCUMENT_TYPE)
+            .name("bar")
+            .build();
+
+    private static final IndexField FIELD_1 = IndexFieldImpl.builder()
+            .fldName("ID")
+            .fldType(FieldType.ID)
+            .analyzerType(AnalyzerType.NUMERIC)
+            .build();
+    private static final IndexField FIELD_2 = IndexFieldImpl.builder()
+            .fldName("Name")
+            .fldType(FieldType.TEXT)
+            .analyzerType(AnalyzerType.ALPHA_NUMERIC)
+            .build();
+    private static final IndexField FIELD_3 = IndexFieldImpl.builder()
+            .fldName("State")
+            .fldType(FieldType.BOOLEAN)
+            .analyzerType(AnalyzerType.KEYWORD)
+            .build();
+
+    @Inject
+    IndexFieldDao indexFieldDao;
+    @Inject
+    IndexDbConnProvider indexDbConnProvider;
+
+    @BeforeEach
+    void setUp() throws SQLException {
+        final Injector injector = Guice.createInjector(
+                new IndexDbModule(),
+                new IndexDaoModule(),
+                new TestModule());
+        injector.injectMembers(this);
+
+        JooqUtil.transaction(indexDbConnProvider, context -> {
+            LOGGER.info("Tear down");
+            context.deleteFrom(INDEX_FIELD).execute();
+            context.deleteFrom(INDEX_FIELD_SOURCE).execute();
+
+            assertThat(count(context, INDEX_FIELD))
+                    .isEqualTo(0);
+            assertThat(count(context, INDEX_FIELD_SOURCE))
+                    .isEqualTo(0);
+        });
+    }
+
+    @Test
+    void addFields() {
+        List<IndexField> fields = getFields(DOC_REF_1);
+
+        assertThat(fields.size())
+                .isEqualTo(0);
+
+        indexFieldDao.addFields(DOC_REF_1, List.of(FIELD_1));
+
+        fields = getFields(DOC_REF_1);
+
+        assertThat(fields.size())
+                .isEqualTo(1);
+
+        // Now add all three fields, so field 1 is ignored
+        indexFieldDao.addFields(DOC_REF_1, List.of(FIELD_1, FIELD_2, FIELD_3));
+
+        fields = getFields(DOC_REF_1);
+
+        assertThat(fields.size())
+                .isEqualTo(3);
+    }
+
+    @Disabled // Verifying lock behaviour
+    @Test
+    void test() throws ExecutionException, InterruptedException {
+        indexFieldDao.addFields(DOC_REF_1, List.of(FIELD_1));
+
+        final CountDownLatch startLatch = new CountDownLatch(1);
+
+        CompletableFuture.runAsync(() -> {
+            JooqUtil.context(indexDbConnProvider, context -> {
+                try {
+                    startLatch.await();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+
+                final Record1<Integer> rec = context.select(INDEX_FIELD_SOURCE.ID)
+                        .from(INDEX_FIELD_SOURCE)
+                        .where(INDEX_FIELD_SOURCE.UUID.eq(DOC_REF_1.getUuid()))
+                        .and(INDEX_FIELD_SOURCE.TYPE.eq(DOC_REF_1.getType()))
+                        .fetchOne();
+
+                final Integer id = rec.get(INDEX_FIELD_SOURCE.ID);
+
+                LOGGER.info("id {}", id);
+            });
+        });
+
+        CompletableFuture.runAsync(() -> {
+            JooqUtil.context(indexDbConnProvider, context -> {
+                context.select(INDEX_FIELD_SOURCE.ID)
+                        .from(INDEX_FIELD_SOURCE)
+                        .where(INDEX_FIELD_SOURCE.UUID.eq(DOC_REF_1.getUuid()))
+                        .and(INDEX_FIELD_SOURCE.TYPE.eq(DOC_REF_1.getType()))
+                        .forUpdate()
+                        .fetch();
+                LOGGER.info("Done lock");
+                startLatch.countDown();
+
+                ThreadUtil.sleepIgnoringInterrupts(5_000);
+            });
+        }).get();
+    }
+
+    /**
+     * Make sure multiple threads can concurrently add different lists of fields
+     */
+    @Test
+    void testMultiThread() throws ExecutionException, InterruptedException {
+        final int threads = 10;
+        final ExecutorService executorService = Executors.newFixedThreadPool(threads);
+        final CountDownLatch startLatch = new CountDownLatch(1);
+
+        final List<CompletableFuture<?>> futures = new ArrayList<>();
+        for (int i = 0; i < threads; i++) {
+            final int finalI = i;
+            final CompletableFuture<Void> future = CompletableFuture.runAsync(
+                    ThrowingRunnable.unchecked(() -> {
+                        final List<IndexField> fields = finalI % 2 == 0
+                                ? List.of(FIELD_1, FIELD_3)
+                                : List.of(FIELD_2, FIELD_3);
+                        startLatch.await();
+                        indexFieldDao.addFields(DOC_REF_1, fields);
+                        LOGGER.debug("Thread {} complete", finalI);
+                    }), executorService);
+            futures.add(future);
+        }
+        startLatch.countDown();
+
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+                .get();
+
+        final List<IndexField> fields = getFields(DOC_REF_1);
+
+        assertThat(fields.size())
+                .isEqualTo(3);
+    }
+
+    @Test
+    void findFields() {
+        List<IndexField> fields = getFields(DOC_REF_1);
+        assertThat(fields.size())
+                .isEqualTo(0);
+
+        indexFieldDao.addFields(DOC_REF_1, List.of(FIELD_1, FIELD_2, FIELD_3));
+        indexFieldDao.addFields(DOC_REF_2, List.of(FIELD_3));
+
+        fields = getFields(DOC_REF_1);
+        assertThat(fields.size())
+                .isEqualTo(3);
+
+        fields = getFields(DOC_REF_2);
+        assertThat(fields.size())
+                .isEqualTo(1);
+    }
+
+    private List<IndexField> getFields(final DocRef docRef) {
+        final ResultPage<IndexField> resultPage = indexFieldDao.findFields(
+                new FindIndexFieldCriteria(
+                        new PageRequest(),
+                        Collections.emptyList(),
+                        docRef,
+                        null));
+        return resultPage.getValues();
+    }
+}

--- a/unreleased_changes/20240410_200322_113__4195.md
+++ b/unreleased_changes/20240410_200322_113__4195.md
@@ -1,0 +1,24 @@
+* Issue **#4195** : Fix SQL deadlock with dynamic index fields.
+
+
+```sh
+# ********************************************************************************
+# Issue title: SQL Deadlock 7.3
+# Issue link:  https://github.com/gchq/stroom/issues/4195
+# ********************************************************************************
+
+# ONLY the top line will be included as a change entry in the CHANGELOG.
+# The entry should be in GitHub flavour markdown and should be written on a SINGLE
+# line with no hard breaks. You can have multiple change files for a single GitHub issue.
+# The  entry should be written in the imperative mood, i.e. 'Fix nasty bug' rather than
+# 'Fixed nasty bug'.
+#
+# Examples of acceptable entries are:
+#
+#
+# * Issue **123** : Fix bug with an associated GitHub issue in this repository
+#
+# * Issue **namespace/other-repo#456** : Fix bug with an associated GitHub issue in another repository
+#
+# * Fix bug with no associated GitHub issue.
+```

